### ISSUE TITLE
Bug: [no-for-in-array] rule does not detect complex non-array situation

### ISSUE THREAD

**Issue by HolgerJeromin** [typescript-eslint/typescript-eslint#10743]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.ts&code=MYGwhgzhAEBmD28A8ARaBTAHgF3QOwBMYJsAnASzwHMBtAXWgB9oAldYeUgpEi6gGmh4ArgFsARulIA%2BadADeAKGjQADqXiqAXNDTMRIEE2jDC6WJXQEA3Mui4S0ALzQAFAEpncpSpUc8jmCkpGAAns72ABbkEAB06pq2vtDksG4AgsFhsTGZIaGuQfnunj7J0AD0FULwALQIpLWUtUVh0KToqpzYMBzB7NggoXbJDW7%2BjuQpeNCtoaUj5RPwIOixIPBUhVmhNOR07knlfvAB2NAA%2BuS4ohFze3SLxypVAHoA-E-QAL6L3xggCDoBRfKo1eqcJp4Fo7aAEeDoGB4eDnDpdUjnPodYCDUKxaAAFWiMBiNVmOy%2BY1cE3OU0o5OKIOe0GWq3Wm22%2BQehy%2BJzOl2u6FuLnu%2B155Ten2Ov2SMp%2Bil%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDMPpbmtAIbRoQwuiiJRfSODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class foo<D extends string[] | Record<string, number>> {
  prop: D | null | undefined;
  test = () => {
    const array = this.prop;
    if (Array.isArray(array)) {
      // no-for-in-array reports correctly
      for (const i in array) {
        console.log(array[i]);
        const _item = array[i]
            //^?
      }
    } else {
      // no-for-in-array does not report correctly. This is no array
      for (const i in array) {
        console.log(array[i]);
        const _item = array[i]
            //^?
      }
    }
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-for-in-array": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

We have a check `Array.isArray()`, but TS does not detect it is no array-like.

### Actual Result

Since #10535 ("report on any type which may be an array or array-like") this code get reported as an error.

I am totally fine to accept that this complex thing is failing the "may be an array" test. 
I know Typescript is [not perfect](https://github.com/microsoft/TypeScript/issues/9998) in flow analytics.

But perhaps you are interested in this.

**Comment by kirkwaiblinger** [comment]
Hi @HolgerJeromin !

It looks like this is specifically an issue to do with the type parameter/generic.... We can simplify the repro to 
```ts
function foo<T extends string[] | Record<string, number>>(array: T) {
  if (Array.isArray(array)) {
    // no-for-in-array reports correctly
    for (const i in array) {
    }
  } else {
    // no-for-in-array does not report correctly. This is no array
    for (const i in array) {
    }
  }  
}
```
Compare to 
```ts
declare const x: string[] | Record<string, number> | null | undefined;
if (Array.isArray(x)) {
  // reports correctly
  for (const i in x) {
  }
} else {
  // no false positive
  for (const i in x) {
  }
}
```

([playground](https://typescript-eslint.io/play/#ts=5.7.3&fileType=.ts&code=GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdESoAnGMAcwG0BdRAH0QCVsI5Ki0LqdADSIwIALYAjbJQB8sgBQBDSpSUBPAFyIMASkQBvAFCJEMYIgUBBVRoB0MEjbXrlt9bv3HTpgPS-ROABaFEogmiCVF0RKbAAHHigyblUuKAAbdRMfUMtuMAozMyQojS9s0wBfbMqcdJJsQwrEf0CQnnCwSPdEIjhsMjA4KBj4xMQU2OhMux0AC0czQbhEUqyfZB48hEKYYtX3co3qqtMjaqMjIi50lUb8wtxtARoGZjZOFL4X4VEJaTkrD%2B6XSQPA12ANGwRAA3EZzJZnPZHEjXLhPE1TLkFA8RnsaIh0U0auc6g1MZtKNsCnj9kTjCTKkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDMPpbmtAIbRoQwuiiJRfSODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

---

We could possibly investigate this further, but I'm not sure whether it'll be feasible to fix on our side, since TS appears to give the type of the `array` in the else branch as `T`, not as `Exclude<T, string[]>` (which the rule [would correctly ignore](https://typescript-eslint.io/play/#ts=5.7.3&showAST=types&fileType=.ts&code=GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdESoAnGMAcwG0BdRAH0QCVsI5Ki0LqdADSIwIALYAjbJQB8sgBQBDSpSUBPAFyIMASkQBvAFCJEMYIgUBBVRoB0MEjbXrlt9bv3HTpgPS-ROABaFEogmiCVF0RKbAAHHigyblUuKAAbdRMfUMtuMAozMyQojS9s0wBfbMqcdJJsQwrEf0CQnnCwSPdEIjhsMjA4KBj4xMQU2OhMux0AC0czQbhEUqyfZB48hEKYYtWepTIAUVwIdJAibEwRcABrIYB3MCZZctaNz6-TAD0Afma1SqpiM1SMRiu5xUjXyhVw2gENAYzDYnBSfERwlEEmkclY2PS6Xx4CuwBo2CIAG4jOZLM57I56a5cJ4mqZcgpYSM9jRECymjVQXUGmzNpRtgVuft%2BcZBZUgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDMPpbmtAIbRoQwuiiJRfSODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)). So I think most likely this will get closed as external/wontfix. But good catch!

FWIW - I would personally recommend just banning `for...in` entirely but that's just me 🙃 (https://github.com/typescript-eslint/typescript-eslint/issues/10558).

**Comment by HolgerJeromin** [comment]
Yeah, TS is quite unsure about isArray on this variable. 
The type in the isArray branch is `T & any[]` which is not exact either.
(lib.es5.d.ts has `isArray(arg: any): arg is any[];` ) There are [some issues](https://github.com/microsoft/TypeScript/issues/17002) in the ts tracker on this topic.

> So I think most likely this will get closed as external/wontfix. But good catch!

As I wrote: fine with me :)

**Comment by bradzacher** [comment]
This is a TS limitation. TS does not refine the generic like you're expecting. In the true branch TS's "refinement" is `D & any[]`. But in the false branch the type is just `D | null | undefined` - it does not refine `D`.

So when we inspect the type we see a plain generic that could be an array.

So this is working as expected.
